src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -63,68 +72,229 @@ void BatchNorm<InputDataType, OutputDataType>::Reset()
+
+  // We will calculate mini batch norm on each channel / feature map.
+  if (!deterministic)
   {
-    mean = arma::mean(input, 1);
-    variance = arma::var(input, 1, 1);
+    Log::Assert(input.n_cols > 0, "Empty input. Batch size must be \
+        greater than zero.");
 
-    // Normalize the input.
-    output = input.each_col() - mean;
-    inputMean = output;
-    output.each_col() /= arma::sqrt(variance + eps);
+    Log::Assert(input.n_rows % size == 0, "Input features must be divisible \
+        by feature maps.");
 
-    // Use Welford method to compute the sample variance and mean.
-    for (size_t i = 0; i < input.n_cols; i++)
+    // Input size refers to inputWidth * inputHeight.
+    // If inputSize equals one, input corresponds to output from linear layer.
+    const size_t inputSize = input.n_rows / size;
+
+    // Linear output.
+    if (inputSize == 1)
     {
-      count += 1;
+      mean = arma::mean(input, 1);
+      variance = arma::var(input, 1, 1);
+
+      // Normalize the input.
+      output = input.each_col() - mean;
+      inputMean = output;
+      output.each_col() /= arma::sqrt(variance + eps);
+
+      // Use Welford method to compute the sample variance and mean.
+      for (size_t i = 0; i < input.n_cols; i++)
+      {
+        count += 1;
+
+        OutputDataType diff = input.col(i) - runningMean;
+        if (average)
+        {
+          runningMean = runningMean + diff / count;
+          runningVariance += diff % (input.col(i) - runningMean);
+        }
+        else
+        {
+          runningMean = (1 - momentum) * runningMean + momentum * mean;
+          runningVariance = (1 - momentum) * runningVariance +
+              momentum * variance;
+        }
+      }
+
+      // Reused in the backward and gradient step.
+      normalized = output;
 
-      OutputDataType diff = input.col(i) - runningMean;
-      runningMean = runningMean + diff / count;
-      runningVariance += diff % (input.col(i) - runningMean);
+      // Scale and shift the output.
+      output.each_col() %= gamma;
+      output.each_col() += beta;
+      return;
     }
 
-    // Reused in the backward and gradient step.
-    normalized = output;
+    // Calculate mean and variance over all channels.
+    mean = OutputDataType(1, size, arma::fill::zeros);
+    variance = OutputDataType(1, size, arma::fill::zeros);
+
+    // Initialize output to same size and values for convience.
+    output = input;
+    normalized.set_size(arma::size(input));
+    inputMean.set_size(arma::size(input));
+    double nElements = input.n_elem / size;
+    nElements = 1.0 * nElements / (1.0 * nElements - 1.0);
+
+    for (size_t channelIdx = 0; channelIdx < size; channelIdx++)
+    {
+      OutputDataType meanTemp, varianceTemp;
+      meanTemp = arma::mean(arma::mean(input(arma::span(
+          channelIdx * inputSize, (channelIdx + 1) * inputSize - 1),
+          arma::span()), 1), 0);
+      mean(channelIdx) = meanTemp(0);
```

Comment:
Do you think it might be better to reshape the input as a cube and then take the mean? 
We could then probably avoid the loop. Let me know what you think.

Reply:
From understanding point of view, that would be better. I think that might also give better speed. So I guess I should change the implementation?

Reply:
Yes sure.

Reply:
Great, I'll make the changes.

Reply:
Hey @saksham189, I have made the changes for forward in a cpp file. All output for both training and testing match with pytorch. The cpp file can be viewed using this [Link](https://github.com/kartikdutt18/mlpack-test/blob/master/batchnorm_3d_test/batchnorm_3d.cpp). I'll make the changes for backward in the next hour or so. Sorry I couldn't make the changes yesterday, I had some errands to run.

Reply:
I have made the changes, I will add more tests (detailed) to show running deterministic output, mean and variance.

Reply:
That sounds good, thanks.

Reply:
I have added a couple of tests for output, variance, deterministic output and mean. Should I also add GradientBatchNormTest with Convolutional layer.

Reply:
Yes I think that might be a good idea.